### PR TITLE
Add general integer exponents

### PR DESCRIPTION
**Pull request by kvhuguenin**
Add exponents >3 to the main branch.
The core changes in the potential have already been implemented.
The most important change that still has to be implemented and tested in the rest of the code is that `exponent` now is of data type `int` rather than `float`.
Also, we need some good unit tests. I was thinking of a combination of regression tests for some of the more complicated structures + generalized Madelung constants that I have computed in the past with a different code.

# Contributor (creator of pull-request) checklist

 - [ ] Tests updated (for new features and bugfixes)?
 - [ ] Documentation updated (for new features)?
 - [ ] Issue referenced (for PRs that solve an issue)?

# Reviewer checklist

 - [ ] CHANGELOG updated with public API or any other important changes?

----
📚 Documentation preview 📚: https://torch-pme--128.org.readthedocs.build/en/128/

### CODE REVIEW

File: src/torchpme/potentials/inversepowerlaw.py [hunk 1]
```diff
@@ -17,6 +17,27 @@ def gamma(x: torch.Tensor) -> torch.Tensor:
     return torch.exp(gammaln(x))
 
 
+# Auxilary function for stable Fourier transform implementation
+def gammainc_upper_over_powerlaw(exponent, zz):
```
Comment (PicoCentauri) [inline src/torchpme/potentials/inversepowerlaw.py:21]: Give the equation in the docstring here.

File: src/torchpme/potentials/inversepowerlaw.py [hunk 2]
```diff
@@ -17,6 +17,27 @@ def gamma(x: torch.Tensor) -> torch.Tensor:
...
+        return 2 * (
+            torch.exp(-zz) - torch.sqrt(torch.pi * zz) * torch.erfc(torch.sqrt(zz))
+        )
+    if exponent == 5:
+        return torch.exp(-zz) + zz * torch.expi(-zz)
+    if exponent == 6:
+        return (
+            (2 - 4 * zz) * torch.exp(-zz)
+            + 4 * torch.sqrt(torch.pi) * zz**1.5 * torch.erfc(torch.sqrt(zz))
+        ) / 3
```
Comment (PicoCentauri) [inline src/torchpme/potentials/inversepowerlaw.py:38]: We need an error message if the exponent is not supported.

I think there is a check already at the init of `InversePowerLawPotential`.

I would maybe call this function in the init of `InversePowerLawPotential` for the given `exponent` and maybe `zz=0` to not have a duplicated test. But up to you.

**Review by PicoCentauri — COMMENTED** [review]
As test I would just do a regression test for the function.

File: src/torchpme/potentials/inversepowerlaw.py [hunk 3]
```diff
@@ -17,6 +17,30 @@ def gamma(x: torch.Tensor) -> torch.Tensor:
...
+def gammainc_upper_over_powerlaw(exponent, zz):
+    if exponent not in [1, 2, 3, 4, 5, 6]:
+        raise ValueError(f"Unsupported exponent: {exponent}")
+
+    if exponent == 1:
+        return torch.exp(-zz) / zz
+    if exponent == 2:
+        return torch.sqrt(torch.pi / zz) * torch.erfc(torch.sqrt(zz))
+    if exponent == 3:
+        return -torch.expi(-zz)
```
Comment (E-Rum) [inline src/torchpme/potentials/inversepowerlaw.py:30]: As far as I know `torch.expi` doesn't exist in `torch` module ...

File: src/torchpme/potentials/inversepowerlaw.py [hunk 4]
```diff
@@ -17,6 +17,30 @@ def gamma(x: torch.Tensor) -> torch.Tensor:
...
+        )
+    if exponent == 5:
+        return torch.exp(-zz) + zz * torch.expi(-zz)
+    if exponent == 6:
+        return (
+            (2 - 4 * zz) * torch.exp(-zz)
+            + 4 * torch.sqrt(torch.pi) * zz**1.5 * torch.erfc(torch.sqrt(zz))
+        ) / 3
+
+
```
Comment (E-Rum) [inline src/torchpme/potentials/inversepowerlaw.py:43]: Can we double-check that everything here is correct? With the new implementation, `test_values_ewald.py` fails for inverse potentials with an exponent of 1, as it does not coincide with the Madelung constant.

**Comment by E-Rum** [comment]
The current status is as follows: 

For arbitrary odd integer exponents ( p >=3 ), we need the exponential integral, which is not implemented in PyTorch. The current solution is to use the existing `SplinedPotential` to model  1/r^p type potentials. However, before fully adopting it, I plan to use the scipy module, which includes the exponential integral, to verify that the values generated by `SplinedPotential` are consistent with the theoretical ones.

However, there is a specific issue when p = 3. In this case, certain variables become degenerate, leading to situations where we raise values to the power of zero and even encounter division by zero (e.g. `background_correction` function).

Can someone provide insights into this issue and suggest potential solutions?

**Comment by E-Rum** [comment]
Here’s the corrected version of your text:

In the last commit, I made the following changes:
I temporarily inserted a `SciPy` function that contains the exponential integral into the existing `InversePowerLawPotential` class. Even though this breaks PyTorch differentiability, it temporarily allows us to have a reference value.

As discussed with @ceriottm, one possible solution to overcome the absence of the exponential integral in PyTorch is to implement it using a spline potential. Therefore, I also drafted how this could be implemented in a separate `integerspline.py` module.

So far, the numerically splined implementation coincides with the reference SciPy one with good accuracy.

However, we still need to discuss what to do with ￼p <= 3. From what I found in the literature, our current implementation is correct for p>3￼. However, for p<=3, one should be careful as certain constraints (e.g., charge or dipole neutrality) need to be enforced (this is apperently why we are having trouble right now in the code with p = 3). Please correct me if I misinterpreted this while reviewing the literature.

If that’s the case, I think we should add some constraints in the code for the above-mentioned exponents.

**Comment by ceriottm** [comment]
> In the last commit, I made the following changes: I temporarily inserted a `SciPy` function that contains the exponential integral into the existing `InversePowerLawPotential` class. Even though this breaks PyTorch differentiability, it temporarily allows us to have a reference value.
> 
> As discussed with @ceriottm, one possible solution to overcome the absence of the exponential integral in PyTorch is to implement it using a spline potential. Therefore, I also drafted how this could be implemented in a separate `integerspline.py` module.
> 
> So far, the numerically splined implementation coincides with the reference SciPy one with good accuracy.
> 
> However, we still need to discuss what to do with ￼p <= 3. From what I found in the literature, our current implementation is correct for p>3￼. However, for p<=3, one should be careful as certain constraints (e.g., charge or dipole neutrality) need to be enforced (this is apperently why we are having trouble right now in the code with p = 3). Please correct me if I misinterpreted this while reviewing the literature.
> 
> If that’s the case, I think we should add some constraints in the code for the above-mentioned exponents.

The idea of the "background correction" is to add a uniform density that cancels out the total pseudocharge making the system overall neutral. I see no reason why this wouldn't be doable for p=2,3 so there might be an error in the expression for p=3

**Comment by E-Rum** [comment]
This is from 
Williams, D. E. (1989). Accelerated Convergence Treatment of R−n Lattice Sums. Crystallography Reviews, 2(1), 3–23. https://doi.org/10.1080/08893118908032944

Screenshot 2025-01-06 at 14 07 56

Thus, there are indeed some restrictions for p<3, which might be causing the divergences. Alternatively, I might have misunderstood the issue, and we may have implemented the `background_correction` incorrectly.

**Comment by E-Rum** [comment]
Okay, implementing a custom backpropagation-compatible PyTorch exponential integral turned out to be not that hard(thanks @kvhuguenin for noticing it), so it’s done. Now, we need to resolve the `background_correction` problem and add tests. I will add tests soon to compare the values and derivatives of the` InversePowerLawPotential` with the splined implementation.

**Comment by E-Rum** [comment]
I added tests to compare theoretical values and their derivatives with the splined potential for exponents 4, 5, and 6. I also decided to delete the SplinedInteger potential, as it has become redundant now that we have theoretical backpropagatable integer potentials. The only thing left to check is the background correction and what’s happening with  p = 3 , and then it will be ready to merge.

File: src/torchpme/potentials/inversepowerlaw.py [hunk 5]
```diff
@@ -17,6 +18,49 @@ def gamma(x: torch.Tensor) -> torch.Tensor:
     return torch.exp(gammaln(x))
 
 
+# Custom exponential function to have an autograd-compatible version of the exponential
+class CustomE1(torch.autograd.Function):
```
Comment (PicoCentauri) [inline src/torchpme/potentials/inversepowerlaw.py:22]: Very cool!

I think you can put the comment as the docstring.

Also, a question. Probably you checked carefully, but does this has to be a class or can it be also a function? If there is no way we can also wrap 
`CustomE1.apply(z)` into a function `exp1`.

and if this is a public function we should also add a simple test...

File: src/torchpme/potentials/inversepowerlaw.py [hunk 6]
```diff
@@ -17,6 +18,49 @@ def gamma(x: torch.Tensor) -> torch.Tensor:
...
+
+    @staticmethod
+    def forward(ctx, input):
+        ctx.save_for_backward(input)
+        if input.is_cuda:
+            input_cpu = input.cpu()
+            result = torch.tensor(exp1(input_cpu.numpy()), device=input.device)
+        else:
+            result = torch.tensor(exp1(input.numpy()))
+        return result
```
Comment (PicoCentauri) [inline src/torchpme/potentials/inversepowerlaw.py:33]: Maybe this is a bit simpler?

```suggestion
    @staticmethod
    def forward(ctx, input):
        ctx.save_for_backward(input)
        if not input.is_cpu:
            input_numpy = input.cpu().numpy()
        else:
            input_numpy = input.numpy()

        return torch.tensor(exp1(input_numpy), device=input.device, dtype=input.dtype)
```

**Review by PicoCentauri — COMMENTED** [review]
Awesome!

File: src/torchpme/potentials/inversepowerlaw.py [hunk 7]
```diff
@@ -17,6 +18,48 @@ def gamma(x: torch.Tensor) -> torch.Tensor:
...
+        (input,) = ctx.saved_tensors
+        return -grad_output * torch.exp(-input) / input
+
+
+def torch_exp1(input):
+    """Wrapper for the custom exponential integral function."""
+    return CustomExp1.apply(input)
+
+
+def gammaincc_over_powerlaw(exponent: torch.Tensor, z: torch.Tensor) -> torch.Tensor:
```
Comment (PicoCentauri) [inline src/torchpme/potentials/inversepowerlaw.py:41]: Should we maybe move these to a `lib.math.py`?

Just a suggestion though

**Review by PicoCentauri — COMMENTED** [review]
Very nice. I have just a suggestion to keep the repo a bit clean.

Thanks @E-Rum !

**Review by PicoCentauri — APPROVED** [review]
